src/client/Microsoft.Identity.Client/ManagedIdentity/ImdsV2ManagedIdentitySource.cs

```diff
@@ -0,0 +1,185 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Http.Retry;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    internal class ImdsV2ManagedIdentitySource : AbstractManagedIdentity
+    {
+        private const string CsrMetadataPath = "/metadata/identity/getPlatformMetadata";
+
+        private static CsrMetadata _csrMetadata;
+
+        public static async Task<bool> GetCsrMetadataAsync(RequestContext requestContext)
+        {
+            string uami = null;
+            switch (requestContext.ServiceBundle.Config.ManagedIdentityId.IdType)
+            {
+                case AppConfig.ManagedIdentityIdType.ClientId:
+                    uami = requestContext.ServiceBundle.Config.ManagedIdentityId.UserAssignedId;
+                    requestContext.Logger.Info("[Managed Identity] ImdsV2 supports user-assigned client id.");
+                    break;
+
+                case AppConfig.ManagedIdentityIdType.ResourceId:
+                    requestContext.Logger.Info("[Managed Identity] ImdsV2 doesn't support user-assigned resource id. Either provide a client id, or use a system-assigned managed identity.");
+                    return false;
+
+                case AppConfig.ManagedIdentityIdType.ObjectId:
+                    requestContext.Logger.Info("[Managed Identity] ImdsV2 doesn't support user-assigned object id. Either provide a client id, or use a system-assigned managed identity.");
+                    return false;
+
+                default:
+                    requestContext.Logger.Info("[Managed Identity] ImdsV2 supports system-assigned managed identity.");
+                    break;
+            }
+
+            string queryParams = $"api-version={ImdsManagedIdentitySource.ImdsApiVersion}";
+            if (!string.IsNullOrEmpty(uami))
+            {
+                queryParams += $"&uaid={uami}";
+            }
+
+            Uri csrMetadataEndpoint = ImdsManagedIdentitySource.GetValidatedEndpoint(requestContext.Logger, CsrMetadataPath, queryParams);
+
+            var headers = new Dictionary<string, string>
+            {
+                { "Metadata", "true" },
+                { "x-ms-client-request-id", Guid.NewGuid().ToString() }
```

Comment:
correlation ID is from the request context.

Reply:
done